crates/consensus/beacon/src/engine/hooks/mod.rs

```diff
@@ -0,0 +1,116 @@
+use reth_interfaces::sync::SyncState;
+use reth_primitives::BlockNumber;
+use std::task::{Context, Poll};
+
+mod controller;
+pub(crate) use controller::HooksController;
+
+mod prune;
+pub use prune::PruneHook;
+
+/// Collection of [hooks][`Hook`].
+#[derive(Default)]
+pub struct Hooks {
+    inner: Vec<Box<dyn Hook>>,
+}
+
+impl Hooks {
+    /// Creates a new empty collection of [hooks][`Hook`].
+    pub fn new() -> Self {
+        Self { inner: Vec::new() }
+    }
+
+    /// Adds a new [hook][`Hook`] to the collection.
+    pub fn add<H: Hook>(&mut self, hook: H) {
+        self.inner.push(Box::new(hook))
+    }
+}
+
+/// Hook that will be run during the main loop of
+/// [consensus engine][`crate::engine::BeaconConsensusEngine`].
+pub trait Hook: Send + Sync + 'static {
```

Comment:
let's also make this `fmt::Debug`

Reply:
not sure we want this, because otherwise we'll need to constraint the `DB` generic type parameter in pruner with `Debug` too. I removed the custom `Debug` implementation instead, so now we just use the `Hook::name` when logging.